### PR TITLE
autoTypings.json changes

### PR DESCRIPTION
**Pull request by jramsay**

### CODE REVIEW

File: src/services/jsTyping.ts [hunk 1]
```diff
@@ -66,20 +107,25 @@ namespace ts.JsTyping {
...
             try {
-                let cacheTsdJsonDict = tsdJsonCacheInfo.value;
-                // The "notFound" property in the tsd.json is a list of items that were not found in DefinitelyTyped. 
-                // Therefore if they don't come with d.ts files, we should not retry downloading these packages.
+                // The "notFound" property in the tsd.json is a list of items that were not found in DefinitelyTyped.
+                // Therefore if they don't come with d.ts files we should not retry downloading these packages.
                 if (cacheTsdJsonDict.hasOwnProperty("notFound")) {
                     for (let notFoundTypingName of cacheTsdJsonDict["notFound"]) {
-                        if (inferredTypings.hasOwnProperty(notFoundTypingName) && !inferredTypings[notFoundTypingName]) {
+                        if (inferredTypings.hasOwnProperty(notFoundTypingName) && typeof inferredTypings[notFoundTypingName] === 'string') {
```
Comment (zhengbli) [inline src/services/jsTyping.ts]: Validate again

File: src/services/jsTyping.ts [hunk 2]
```diff
@@ -6,40 +6,81 @@
...
 
+    type AutoTypingsInfo = {
+        _autoTypings: string[],
+        autoTypingsEnabled: boolean,
+        include: string[],
+        exclude: string[]
+    };
+
     type HostType = {
+        directoryExists: (fileName: string) => boolean;
```
Comment (zhengbli) [inline src/services/jsTyping.ts]: use `path` instead of `fileName` as it is a directory

File: src/services/jsTyping.ts [hunk 3]
```diff
@@ -6,40 +6,81 @@
...
     var inferredTypings: Map<string> = {};
 
-    function tryParseJson(jsonPath: string) {
-        try {
-            return JSON.parse(jsonPath);
+    function tryParseJson(host: HostType, jsonPath: string): any {
+        if (host.fileExists(jsonPath)) {
+            try {
+                // Remove whitespace and strip out single-line comments
+                let contents = host.readFile(jsonPath).replace(/\/\/(.*)(\r*)(\n*)|\r?\n|\r|/g, "");
```
Comment (zhengbli) [inline src/services/jsTyping.ts]: Is it the same with `/\/\/(.*)(\r*)(\n*)/g`? 
Also, it doesn't seem to remove whitespaces? Although I don't think we need to

File: src/services/jsTyping.ts [hunk 4]
```diff
@@ -6,40 +6,81 @@
...
-        catch (e) { return undefined; }
+        return undefined;
     }
     
     /**
      * @param cachePath is the path to the cache location, which contains a tsd.json file and a typings folder
      */
     export function discoverTypings(
-        host: HostType, fileNames: string[], cachePath: string, compilerOptions?: CompilerOptions, safeList?: string[], noDevDependencies?: boolean)
+        host: HostType, fileNames: string[], cachePath: string, typingsConfigPath: string, compilerOptions?: CompilerOptions, safeList?: string[], noDevDependencies?: boolean)
```
Comment (zhengbli) [inline src/services/jsTyping.ts]: Can we call it `projectRootPath` instead

File: src/services/jsTyping.ts [hunk 5]
```diff
@@ -6,40 +6,81 @@
...
         // Directories to search for package.json, bower.json and other typing information
         let searchDirs: string[] = [];
 
+        // Check the typings config path to see if auto-typings is
+        // enabled and if the there are any typings in the include/exclude list
+        let exclude: string[] = [];
+        let autoTypingsJsonPath = ts.combinePaths(ts.normalizePath(typingsConfigPath), "autoTypings.json");
+        let autoTypingsJsonDict = tryParseJson(host, autoTypingsJsonPath);
+        if (autoTypingsJsonDict) {
+            if (autoTypingsJsonDict.hasOwnProperty("autoTypingsEnabled") && !autoTypingsJsonDict["autoTypingsEnabled"]) {
```
Comment (zhengbli) [inline src/services/jsTyping.ts]: `autoTypingJsonDict["autoTypingEnabled"] === false`

Comment (zhengbli) [inline src/services/jsTyping.ts]: Also this seems to ignore `include` and `exclude` if `autoTypingEnabled` is false, is that what we want?

File: src/services/jsTyping.ts [hunk 6]
```diff
@@ -6,40 +6,81 @@
...
+`;
     
     // a typing name to typing file path mapping
     var inferredTypings: Map<string> = {};
 
-    function tryParseJson(jsonPath: string) {
-        try {
-            return JSON.parse(jsonPath);
+    function tryParseJson(host: HostType, jsonPath: string): any {
+        if (host.fileExists(jsonPath)) {
```
Comment (zhengbli) [inline src/services/jsTyping.ts]: As the `_host` has a global references, no need to pass it as a parameter anymore

File: src/services/jsTyping.ts [hunk 7]
```diff
@@ -217,34 +261,79 @@ namespace ts.JsTyping {
...
-        if (host.fileExists(cacheTsdJsonPath)) {
-            let value = tryParseJson(host.readFile(cacheTsdJsonPath));
-            if (value) {
-                return { path: cacheTsdJsonPath, value };
+        // Update autoTypings.json
+        let newAutoTypingsJson: string = autoTypingsComment;
+
+        // Get the list of new and cached injected typings
+        let installedTypingNames: string[] = [];
+        if (installedTypingsCache) {
```
Comment (zhengbli) [inline src/services/jsTyping.ts:296]: `installedKeys` to be consistent with the `isinstalled` function

File: src/services/jsTyping.ts [hunk 8]
```diff
@@ -217,34 +261,79 @@ namespace ts.JsTyping {
...
+            installedTypingNames = ts.filter(newTypingNames, i => isInstalled(i, installedTypingsCache));
+        }
+
+        let cachedTypingNames: string[] = [];
+        for (let cachedTypingPath of cachedTypingsPaths) {
+            cachedTypingNames.push(ts.removeFileExtension(ts.getBaseFileName(cachedTypingPath)));
```
Comment (zhengbli) [inline src/services/jsTyping.ts]: Use folder name instead

File: src/services/jsTyping.ts (hunk 2 above)
Comment (jramsay) [inline src/services/jsTyping.ts]: good point

File: src/services/jsTyping.ts (hunk 4 above)
Comment (jramsay) [inline src/services/jsTyping.ts]: yes

File: src/services/jsTyping.ts (hunk 6 above)
Comment (jramsay) [inline src/services/jsTyping.ts]: done

File: src/services/jsTyping.ts (hunk 7 above)
Comment (jramsay) [inline src/services/jsTyping.ts:296]: done

File: src/services/jsTyping.ts (hunk 5 above)
Comment (jramsay) [inline src/services/jsTyping.ts]: - done.  
- I think we should be OK on the include/exclude front. When autoTypings is disabled we don't try to auto-reference typings.

**Pull request closed without merging**